sopel/plugin.py

```diff
@@ -820,31 +823,168 @@ def add_attribute(function):
 
     How often a function can be triggered on a per-user basis, in a channel,
     or across the server (bot) can be controlled with this decorator. A value
     of ``0`` means no limit. If a function is given a rate of 20, that
     function may only be used once every 20 seconds in the scope corresponding
-    to the parameter. Users on the admin list in Sopel’s configuration are
-    exempted from rate limits.
+    to the parameter::
+
+        from sopel import plugin
+
+        @plugin.rate(10)
+            # won't trigger if used more than once per 10s by a user
+
+        @plugin.rate(10, 10)
+            # won't trigger if used more than once per 10s by a user/channel
+
+        @plugin.rate(10, 10, 2)
+            # won't trigger if used more than once per 10s by a user/channel
+            # and never more than once every 2s
+
+    If a ``message`` is provided, it will be used as the default message sent
+    as a ``NOTICE`` to the user who hit the rate limit::
+
+        @rate(10, 10, 10, message='Hit the rate limit for this function.')
+            # will send a NOTICE
 
     Rate-limited functions that use scheduled future commands should import
     :class:`threading.Timer` instead of :mod:`sched`, or rate limiting will
     not work properly.
+
+    .. versionchanged:: 8.0
+
+        Optional keyword argument ``message`` was added in Sopel 8.
+
+    .. note::
+
+        Users on the admin list in Sopel’s configuration are exempted from rate
+        limits.
+
+    .. seealso::
+
+        You can control each rate limit separately, with their own custom
+        message using :func:`rate_user`, :func:`rate_channel`, or
+        :func:`rate_global`.
+
+    """
+    def add_attribute(function):
+        if not hasattr(function, 'rate'):
+            function.rate = user
+        if not hasattr(function, 'channel_rate'):
+            function.channel_rate = channel
+        if not hasattr(function, 'global_rate'):
+            function.global_rate = server
+        function.default_rate_message = message
+        return function
+    return add_attribute
+
+
+def rate_user(
+    rate: int,
+    message: typing.Optional[str] = None,
+) -> typing.Callable:
+    """Decorate a function to be rate-limited for a user.
+
+    :param rate: seconds between permitted calls of this function by the same
+                 user
+    :param message: optional; message sent as NOTICE when a user hits the limit
+
+    This decorator can be used alone or with the :func:`rate` decorator, as it
+    will always take precedence::
+
+        @rate(10, 10, 10)
+        @rate_user(20, 'You hit your rate limit for this function.')
+            # user limit will be set to 20, other to 10
+            # will send a NOTICE only when a user hits their own limit
+            # as other rate limits don't have any message set
+
+    If you don't provide a message, the default message set (if any) by
+    :func:`rate` will be used instead.
```

Comment:
`rate_channel`'s format wins.

```suggestion
    If you don't provide a message, the default message set by :func:`rate`
    (if any) will be used instead.
```